File: ios/USTAQuery/ViewModels/TournamentDetailViewModel.swift
import SwiftUI

// MARK: - Display Row Types

enum StatusGroup: String, CaseIterable {
    case accepted, alternate, withdrawn, other

    var label: String {
        switch self {
        case .accepted: return "Acceptance"
        case .alternate: return "Alternates"
        case .withdrawn: return "Withdrawn"
        case .other: return "Other"
        }
    }

    var color: Color {
        switch self {
        case .accepted: return AppTheme.tennisGreen
        case .alternate: return .orange
        case .withdrawn: return AppTheme.lossRed
        case .other: return .secondary
        }
    }
}

struct DisplayRow: Identifiable {
    let id = UUID()
    var entries: [TournamentEntry]
    var eventId: String
    var entryPosition: Int?
    var entryStatus: String?
}

// MARK: - ViewModel

@Observable
final class TournamentDetailViewModel {
    let tournamentId: Int
    var tournament: Tournament?
    var entries: [TournamentEntry] = []
    var selectedEventId: String?
    var isLoading = false
    var entriesLoading = false
    var error: String?

    init(tournamentId: Int) {
        self.tournamentId = tournamentId
    }

    func loadTournament() async {
        isLoading = true
        error = nil
        do {
            tournament = try await APIClient.getTournament(id: tournamentId)
            await loadEntries()
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func loadEntries() async {
        entriesLoading = true
        do {
            entries = try await APIClient.getTournamentEntries(id: tournamentId, eventId: selectedEventId)
        } catch {
            entries = []
        }
        entriesLoading = false
    }

    func selectEvent(_ eventId: String?) async {
        selectedEventId = eventId
        await loadEntries()
    }

    // MARK: - Build Display Rows (ported from TournamentDetailPage.tsx)

    var displayRows: [DisplayRow] {
        buildDisplayRows(entries)
    }

    var groupedByStatus: [(StatusGroup, [DisplayRow])] {
        let rows = displayRows
        var groups: [StatusGroup: [DisplayRow]] = [:]
        for row in rows {
            let g = classifyStatus(row.entryStatus)
            groups[g, default: []].append(row)
        }
        return StatusGroup.allCases.compactMap { g in
            guard let rows = groups[g], !rows.isEmpty else { return nil }
            return (g, rows)
        }
    }

    private func classifyStatus(_ status: String?) -> StatusGroup {
        guard let s = status?.uppercased() else { return .other }
        if s.contains("DIRECT") || s == "REGISTERED" { return .accepted }
        if s.contains("ALTERNATE") { return .alternate }
        if s.contains("WITHDRAWN") { return .withdrawn }
        return .other
    }

    private func buildDisplayRows(_ entries: [TournamentEntry]) -> [DisplayRow] {
        var rows: [DisplayRow] = []
        var teamEntries: [TournamentEntry] = []
        var individualsByKey: [String: [TournamentEntry]] = [:]
        var singlesOrOther: [TournamentEntry] = []

        // Classify entries
        for e in entries {
            let isDoubles = (e.eventType ?? "").uppercased().contains("DOUBLES")
            let isTeam = (e.playerName ?? "").contains("/") && (e.firstName?.trimmingCharacters(in: .whitespaces).isEmpty ?? true)

            if isDoubles && isTeam {
                teamEntries.append(e)
            } else if isDoubles && !(e.firstName?.trimmingCharacters(in: .whitespaces).isEmpty ?? true) {
                let ln = (e.lastName ?? "").trimmingCharacters(in: .whitespaces).lowercased()
                let key = "\(e.eventId)::\(e.drawId ?? "")::\(ln)"
                individualsByKey[key, default: []].append(e)
            } else {
                singlesOrOther.append(e)
            }
        }

        // Pair doubles using team entries as lookup
        var pairedIds: Set<String> = []

        for team in teamEntries {
            let names = (team.playerName ?? "").split(separator: "/").map { $0.trimmingCharacters(in: .whitespaces).lowercased() }
            guard names.count == 2 else { continue }

            let drawId = team.drawId ?? ""
            var pair: [TournamentEntry] = []

            for ln in names {
                let key = "\(team.eventId)::\(drawId)::\(ln)"
                if let candidates = individualsByKey[key] {
                    let unused = candidates.filter { !pairedIds.contains($0.participantId ?? "") }
                    let pick = unused.first { $0.entryPosition == team.entryPosition } ?? unused.first
                    if let pick {
                        pair.append(pick)
                        if let pid = pick.participantId { pairedIds.insert(pid) }
                    }
                }
            }

            if pair.count == 2 {
                rows.append(DisplayRow(entries: pair, eventId: team.eventId, entryPosition: team.entryPosition, entryStatus: team.entryStatus))
            } else {
                rows.append(DisplayRow(entries: [team], eventId: team.eventId, entryPosition: team.entryPosition, entryStatus: team.entryStatus))
            }
        }

        // Add unmatched individual doubles entries
        for list in individualsByKey.values {
            for e in list {
                if !pairedIds.contains(e.participantId ?? "") {
                    rows.append(DisplayRow(entries: [e], eventId: e.eventId, entryPosition: e.entryPosition, entryStatus: e.entryStatus))
                }
            }
        }

        // Add singles entries
        for e in singlesOrOther {
            rows.append(DisplayRow(entries: [e], eventId: e.eventId, entryPosition: e.entryPosition, entryStatus: e.entryStatus))
        }

        // Sort by entryPosition when available, otherwise preserve order
        let hasPositions = rows.contains { $0.entryPosition != nil }
        if hasPositions {
            rows.sort { a, b in
                (a.entryPosition ?? Int.max) < (b.entryPosition ?? Int.max)
            }
        }

        return rows
    }
}

File: android/app/src/main/java/com/usta/query/viewmodel/TournamentDetailViewModel.kt
package com.usta.query.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.usta.query.data.api.ApiClient
import com.usta.query.data.model.Tournament
import com.usta.query.data.model.TournamentEntry
import com.usta.query.data.model.TournamentEvent
import kotlinx.coroutines.launch

data class DisplayRow(
    val entries: List<TournamentEntry>,
    val eventId: String,
    val entryPosition: Int?,
    val entryStatus: String?
) {
    val totalPoints: Int get() = entries.sumOf { it.rankingPoints ?: 0 }
    val isPair: Boolean get() = entries.size == 2
}

class TournamentDetailViewModel(private val tournamentId: Int) : ViewModel() {
    var tournament by mutableStateOf<Tournament?>(null)
    var entries by mutableStateOf<List<TournamentEntry>>(emptyList())
    var selectedEventId by mutableStateOf<String?>(null)
    var isLoading by mutableStateOf(false)
    var entriesLoading by mutableStateOf(false)
    var error by mutableStateOf<String?>(null)

    fun loadTournament() {
        isLoading = true
        viewModelScope.launch {
            try {
                tournament = ApiClient.api.getTournament(tournamentId)
                loadEntries()
            } catch (e: Exception) {
                error = e.message
            } finally {
                isLoading = false
            }
        }
    }

    fun loadEntries() {
        entriesLoading = true
        viewModelScope.launch {
            try {
                entries = ApiClient.api.getTournamentEntries(tournamentId, selectedEventId)
            } catch (_: Exception) {
                entries = emptyList()
            } finally {
                entriesLoading = false
            }
        }
    }

    fun selectEvent(eventId: String?) {
        selectedEventId = eventId
        loadEntries()
    }

    fun getDisplayRows(): List<DisplayRow> {
        val rows = mutableListOf<DisplayRow>()
        val teamEntries = mutableListOf<TournamentEntry>()
        val individualsByKey = mutableMapOf<String, MutableList<TournamentEntry>>()
        val singlesOrOther = mutableListOf<TournamentEntry>()

        for (e in entries) {
            val evType = (e.eventType ?: "").uppercase()
            val isDoubles = evType.contains("DOUBLES")
            val isTeamEvent = evType.contains("TEAM")
            val hasSlash = (e.playerName ?: "").contains("/")
            val hasFirstName = (e.firstName ?: "").isNotBlank()

            when {
                isTeamEvent && !hasFirstName -> {
                    // TEAM event synthetic entry: pair summary or team name — skip both
                }
                isDoubles && hasSlash && !hasFirstName -> {
                    // Doubles pair summary (e.g. "Racic/Yamamoto")
                    teamEntries.add(e)
                }
                isDoubles && hasFirstName -> {
                    val ln = (e.lastName ?: "").trim().lowercase()
                    val key = "${e.eventId}::${e.drawId ?: ""}::$ln"
                    individualsByKey.getOrPut(key) { mutableListOf() }.add(e)
                }
                else -> singlesOrOther.add(e)
            }
        }

        val pairedIds = mutableSetOf<String>()

        for (team in teamEntries) {
            val names = (team.playerName ?: "").split("/").map { it.trim().lowercase() }
            if (names.size != 2) continue

            val drawId = team.drawId ?: ""
            val pair = mutableListOf<TournamentEntry>()

            for (ln in names) {
                val key = "${team.eventId}::${drawId}::$ln"
                val candidates = individualsByKey[key]
                if (candidates != null) {
                    val unused = candidates.filter { pairedIds.isEmpty() || !pairedIds.contains(it.participantId ?: "") }
                    val pick = unused.find { it.entryPosition == team.entryPosition } ?: unused.firstOrNull()
                    if (pick != null) {
                        pair.add(pick)
                        pick.participantId?.let { pairedIds.add(it) }
                    }
                }
            }

            if (pair.size == 2) {
                rows.add(DisplayRow(pair, team.eventId, team.entryPosition, team.entryStatus))
            } else {
                rows.add(DisplayRow(listOf(team), team.eventId, team.entryPosition, team.entryStatus))
            }
        }

        for (list in individualsByKey.values) {
            for (e in list) {
                if (!pairedIds.contains(e.participantId ?: "")) {
                    rows.add(DisplayRow(listOf(e), e.eventId, e.entryPosition, e.entryStatus))
                }
            }
        }

        for (e in singlesOrOther) {
            rows.add(DisplayRow(listOf(e), e.eventId, e.entryPosition, e.entryStatus))
        }

        rows.sortWith(compareBy<DisplayRow> { it.entryPosition == null }
            .thenBy { it.entryPosition ?: Int.MAX_VALUE }
            .thenByDescending { it.totalPoints })

        return rows
    }

    fun getGroupedDisplayRows(): Map<String, List<DisplayRow>> {
        return getDisplayRows().groupBy { classifyStatus(it.entryStatus) }
    }

    private fun classifyStatus(status: String?): String {
        val s = status?.uppercase() ?: return "Other"
        return when {
            s.contains("DIRECT") || s == "REGISTERED" -> "Acceptance"
            s.contains("ALTERNATE") || s.contains("UNGROUPED") -> "Alternates"
            s.contains("WITHDRAWN") -> "Withdrawn"
            else -> "Other"
        }
    }
}
